fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/codec/LuceneOptimizedFieldInfosFormat.java

```diff
@@ -20,28 +20,67 @@
 
 package com.apple.foundationdb.record.lucene.codec;
 
+import com.apple.foundationdb.record.lucene.directory.FDBDirectory;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import org.apache.commons.lang3.ArrayUtils;
+import org.apache.commons.lang3.tuple.Pair;
 import org.apache.lucene.codecs.FieldInfosFormat;
 import org.apache.lucene.index.FieldInfos;
+import org.apache.lucene.index.IndexFileNames;
 import org.apache.lucene.index.SegmentInfo;
 import org.apache.lucene.store.Directory;
 import org.apache.lucene.store.IOContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import java.io.IOException;
+import java.util.BitSet;
+import java.util.List;
 
 /**
  * This class attempts to reduce the number of .fnm entries when numerous segments are present in FDB.
  *
  */
 public class LuceneOptimizedFieldInfosFormat extends FieldInfosFormat {
+    public static final int DEFAULT_MAXIMUM_FIELD_INFO_CACHE_SIZE = 128;
+    private static final Logger LOGGER = LoggerFactory.getLogger(LuceneOptimizedFieldInfosFormat.class);
 
     private final FieldInfosFormat fieldInfosFormat;
 
+    private final Cache<Pair<Directory, BitSet>, FieldInfos> fieldInfosCache;
+
     public LuceneOptimizedFieldInfosFormat(FieldInfosFormat fieldInfosFormat) {
         this.fieldInfosFormat = fieldInfosFormat;
+        this.fieldInfosCache = CacheBuilder.newBuilder()
+                .maximumSize(DEFAULT_MAXIMUM_FIELD_INFO_CACHE_SIZE)
+                .build();
     }
 
     @Override
+    @SuppressWarnings("PMD.CloseResource")
     public FieldInfos read(final Directory directory, final SegmentInfo segmentInfo, final String segmentSuffix, final IOContext iocontext) throws IOException {
+        // UnWrap Directory in attempt to use a cache mechanism
+        if (directory instanceof LuceneOptimizedCompoundReader) {
+            Directory next = ((LuceneOptimizedCompoundReader) directory).getDirectory();
+            if (next instanceof LuceneOptimizedWrappedDirectory) {
+                FDBDirectory fdbDirectory = ((LuceneOptimizedWrappedDirectory) next).getFdbDirectory();
+                String fileName = IndexFileNames.segmentFileName(segmentInfo.name, segmentSuffix, "fnm");
+                List<Long> bitSetWords = fdbDirectory.getFDBLuceneFileReference(
+                        LuceneOptimizedWrappedDirectory.convertToDataFile(fileName)).getBitSetWords();
+                BitSet bitSet = BitSet.valueOf(ArrayUtils.toPrimitive(bitSetWords.toArray(new Long[0])));
+                return fieldInfosCache.asMap().computeIfAbsent(Pair.of(fdbDirectory, bitSet), ignore -> {
```

Comment:
Isn't it easier to call
```
fieldsInfoCache.get(Pair.of(fdbDirectory, bitSet), ignore -> {...})
```

Reply:
Need null checks...

Reply:
Thus need the map view which will also perform null checks.